causalml/inference/nn/utils.py

```diff
@@ -0,0 +1,161 @@
+                                   Each row in concat_pred is comprised of (y0, y1, propensity, epsilon)
+    Returns:
+        - (float): aggregated regression + classification loss
+    """
+    return regression_loss(concat_true, concat_pred) + binary_classification_loss(concat_true, concat_pred)
+
+
+def treatment_accuracy(concat_true, concat_pred):
+    """
+    Returns keras' binary_accuracy between treatment and prediction of propensity.
+
+    Args:
+        - concat_true (tf.tensor): tensor of true samples, with shape (n_samples, 2)
+                                   Each row in concat_true is comprised of (y, treatment)
+        - concat_pred (tf.tensor): tensor of predictions, with shape (n_samples, 4)
+                                   Each row in concat_pred is comprised of (y0, y1, propensity, epsilon)
+    Returns:
+        - (float): binary accuracy
+    """
+    t_true = concat_true[:, 1]
+    t_pred = concat_pred[:, 2]
+    return binary_accuracy(t_true, t_pred)
+
+
+def track_epsilon(concat_true, concat_pred):
+    """
+    Tracks the mean absolute value of epsilon.
+
+    Args:
+        - concat_true (tf.tensor): tensor of true samples, with shape (n_samples, 2)
+                                   Each row in concat_true is comprised of (y, treatment)
+        - concat_pred (tf.tensor): tensor of predictions, with shape (n_samples, 4)
+                                   Each row in concat_pred is comprised of (y0, y1, propensity, epsilon)
+    Returns:
+        - (float): mean absolute value of epsilon
+    """
+    epsilons = concat_pred[:, 3]
+    return tf.abs(tf.reduce_mean(epsilons))
+
+
+def make_tarreg_loss(ratio=1., dragonnet_loss=dragonnet_loss_binarycross):
+    """
+    Given a specified loss function, returns the same loss function with targeted regularization.
+
+    Args:
+        ratio (float): weight assigned to the targeted regularization loss component
+        dragonnet_loss (function): a loss function
+    Returns:
+        (function): loss function with targeted regularization, weighted by specified ratio
+    """
+    def tarreg_ATE_unbounded_domain_loss(concat_true, concat_pred):
+        """
+        Returns the loss function (specified in outer function) with targeted regularization.
+        """
+        vanilla_loss = dragonnet_loss(concat_true, concat_pred)
+
+        y_true = concat_true[:, 0]
+        t_true = concat_true[:, 1]
+
+        y0_pred = concat_pred[:, 0]
+        y1_pred = concat_pred[:, 1]
+        t_pred = concat_pred[:, 2]
+
+        epsilons = concat_pred[:, 3]
+        t_pred = (t_pred + 0.01) / 1.02
+        # t_pred = tf.clip_by_value(t_pred,0.01, 0.99,name='t_pred')
+
+        y_pred = t_true * y1_pred + (1 - t_true) * y0_pred
+
+        h = t_true / t_pred - (1 - t_true) / (1 - t_pred)
+
+        y_pert = y_pred + epsilons * h
+        targeted_regularization = tf.reduce_sum(tf.square(y_true - y_pert))
```

Comment:
@jeongyoonlee epsilon is used here. y_pert is an adjusted version of y_pred (adjusted by epsilon times h, which is the targeted regularization part). So epsilon's just another hyperparameter to be tuned it looks like.